### PR TITLE
[SCB-2354]Provide a flexible discovery filter based on instance property and filtering with priority

### PR DESCRIPTION
**Pull request by etteeee**
Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/SCB) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [ ] Each commit in the pull request should have a meaningful subject line and body.
 - [ ] Format the pull request title like `[SCB-XXX] Fixes bug in ApproximateQuantiles`, where you replace `SCB-XXX` with the appropriate JIRA issue.
 - [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [ ] Run `mvn clean install -Pit` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

File: handlers/handler-loadbalance/src/main/java/org/apache/servicecomb/loadbalance/filter/PriorityInstancePropertyDiscoveryFilter.java [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+import org.apache.servicecomb.registry.discovery.DiscoveryTreeNode;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.netflix.config.DynamicPropertyFactory;
+
+/**
+ * Instance property with priority filter
+ */
+public class PriorityInstancePropertyDiscoveryFilter extends AbstractDiscoveryFilter {
```
Comment (liubao68) [inline handlers/handler-loadbalance/src/main/java/org/apache/servicecomb/loadbalance/filter/PriorityInstancePropertyDiscoveryFilter.java:42]: SPI definition is missed. see  META-INF/services/org.apache.servicecomb.registry.discovery.DiscoveryFilter file

File: handlers/handler-loadbalance/src/main/java/org/apache/servicecomb/loadbalance/filter/PriorityInstancePropertyDiscoveryFilter.java [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+      while (!parent.children().containsKey(currentProperty.getPropertyValue())
+          && currentProperty.hasChildren()) {
+        currentProperty = currentProperty.child();
+      }
+    } else {
+      if (currentProperty.hasChildren()) {
+        currentProperty = currentProperty.child();
+      }
+    }
+    LOGGER.info("Discovery instance filter by {}", currentProperty.toString());
```
Comment (liubao68) [inline handlers/handler-loadbalance/src/main/java/org/apache/servicecomb/loadbalance/filter/PriorityInstancePropertyDiscoveryFilter.java:96]: Maybe change to debug is better, or will print a log for each invocation.

File: handlers/handler-loadbalance/src/main/java/org/apache/servicecomb/loadbalance/filter/PriorityInstancePropertyDiscoveryFilter.java [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+      return StringUtils.isEmpty(propertyVal);
+    }
+
+    /**
+     * does property have lower priority children
+     *
+     * @return result
+     */
+    public boolean hasChildren() {
+      return StringUtils.isNotEmpty(propertyVal) && propertyVal.length() < 100;
```
Comment (liubao68) [inline handlers/handler-loadbalance/src/main/java/org/apache/servicecomb/loadbalance/filter/PriorityInstancePropertyDiscoveryFilter.java:166]: What does 100 mean?

File: handlers/handler-loadbalance/src/main/java/org/apache/servicecomb/loadbalance/filter/PriorityInstancePropertyDiscoveryFilter.java (hunk 1 above)
Comment (etteeee) [reply handlers/handler-loadbalance/src/main/java/org/apache/servicecomb/loadbalance/filter/PriorityInstancePropertyDiscoveryFilter.java:42]: fixed

File: handlers/handler-loadbalance/src/main/java/org/apache/servicecomb/loadbalance/filter/PriorityInstancePropertyDiscoveryFilter.java (hunk 2 above)
Comment (etteeee) [reply handlers/handler-loadbalance/src/main/java/org/apache/servicecomb/loadbalance/filter/PriorityInstancePropertyDiscoveryFilter.java]: fixed

File: handlers/handler-loadbalance/src/main/java/org/apache/servicecomb/loadbalance/filter/PriorityInstancePropertyDiscoveryFilter.java (hunk 3 above)
Comment (etteeee) [reply handlers/handler-loadbalance/src/main/java/org/apache/servicecomb/loadbalance/filter/PriorityInstancePropertyDiscoveryFilter.java]: fixed

**Pull request merged**